File: apps/web/lib/business-brain.ts
import "server-only";

import { cookies } from "next/headers";

export const contextSourceTypes = [
  "business_profile",
  "approved_profile",
  "approved_goals",
  "products_services",
  "brand",
  "operating_context",
  "operational_goals",
  "current_tasks",
  "knowledge",
] as const;

export type ContextSourceType = (typeof contextSourceTypes)[number];
export type SourceValidity = "current" | "stale" | "invalidated";
export type SelectionStatus = "included" | "excluded";

export interface ContextSource {
  source_type: ContextSourceType;
  source_reference: string;
  source_version: string;
  authority: string;
  label: string;
  updated_at: string;
  validity: SourceValidity;
  selection_status: SelectionStatus;
  exclusion_reason:
    "not_selected" | "stale" | "invalidated" | "token_budget" | null;
  estimated_tokens: number;
  content_sha256: string;
  content: Record<string, unknown> | null;
}

export interface BusinessContext {
  context_id: string;
  business_id: string;
  purpose: string;
  generated_at: string;
  token_budget: number;
  estimated_tokens: number;
  budget_remaining: number;
  selected_source_types: ContextSourceType[];
  sources: ContextSource[];
  unavailable_sources: Record<string, string>;
  context: string;
  context_sha256: string;
}

function isSource(value: unknown): value is ContextSource {
  if (typeof value !== "object" || value === null) return false;
  const item = value as Partial<ContextSource>;
  return (
    contextSourceTypes.includes(item.source_type as ContextSourceType) &&
    typeof item.source_reference === "string" &&
    typeof item.source_version === "string" &&
    typeof item.authority === "string" &&
    typeof item.label === "string" &&
    typeof item.updated_at === "string" &&
    ["current", "stale", "invalidated"].includes(item.validity ?? "") &&
    ["included", "excluded"].includes(item.selection_status ?? "") &&
    typeof item.estimated_tokens === "number" &&
    typeof item.content_sha256 === "string"
  );
}

function isBusinessContext(value: unknown): value is BusinessContext {
  if (typeof value !== "object" || value === null) return false;
  const item = value as Partial<BusinessContext>;
  return (
    typeof item.context_id === "string" &&
    typeof item.business_id === "string" &&
    typeof item.purpose === "string" &&
    typeof item.generated_at === "string" &&
    typeof item.token_budget === "number" &&
    typeof item.estimated_tokens === "number" &&
    typeof item.budget_remaining === "number" &&
    Array.isArray(item.selected_source_types) &&
    item.selected_source_types.every((source) =>
      contextSourceTypes.includes(source),
    ) &&
    Array.isArray(item.sources) &&
    item.sources.every(isSource) &&
    typeof item.unavailable_sources === "object" &&
    item.unavailable_sources !== null &&
    typeof item.context === "string" &&
    typeof item.context_sha256 === "string"
  );
}

export async function getBusinessContext(options: {
  purpose: string;
  tokenBudget: number;
  sourceTypes: ContextSourceType[];
  knowledgeQuery?: string;
}): Promise<BusinessContext | null> {
  const store = await cookies();
  const session = store.get("id")?.value;
  if (!session) return null;
  const query = new URLSearchParams({
    purpose: options.purpose,
    token_budget: String(options.tokenBudget),
    sources: options.sourceTypes.join(","),
  });
  if (options.knowledgeQuery)
    query.set("knowledge_query", options.knowledgeQuery);
  try {
    const base = process.env.API_INTERNAL_URL ?? "http://localhost:8000";
    const response = await fetch(`${base}/brain/context?${query}`, {
      cache: "no-store",
      headers: { Cookie: `id=${session}` },
      signal: AbortSignal.timeout(5000),
    });
    if (!response.ok) return null;
    const value: unknown = await response.json();
    return isBusinessContext(value) ? value : null;
  } catch {
    return null;
  }
}

File: apps/api/src/foundora/api/business_brain.py
from __future__ import annotations

import re
from datetime import datetime
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query, Response, status
from pydantic import BaseModel

from foundora.api.auth import require_auth
from foundora.auth.service import AuthContext
from foundora.business_brain.service import (
    SOURCE_TYPES,
    BusinessContext,
    ContextBuildRequest,
    ContextService,
    ExclusionReason,
    SelectionStatus,
    SourceType,
    SourceValidity,
)

router = APIRouter(prefix="/brain", tags=["business brain"])
_PURPOSE_PATTERN = re.compile(r"^[a-z][a-z0-9_.-]{0,79}$")


class ContextSourceView(BaseModel):
    source_type: SourceType
    source_reference: str
    source_version: str
    authority: str
    label: str
    updated_at: datetime
    validity: SourceValidity
    selection_status: SelectionStatus
    exclusion_reason: ExclusionReason | None
    estimated_tokens: int
    content_sha256: str
    content: dict[str, object] | None


class BusinessContextView(BaseModel):
    context_id: str
    business_id: str
    purpose: str
    generated_at: datetime
    token_budget: int
    estimated_tokens: int
    budget_remaining: int
    selected_source_types: list[SourceType]
    sources: list[ContextSourceView]
    unavailable_sources: dict[str, str]
    context: str
    context_sha256: str


def _source_types(value: str | None) -> frozenset[SourceType]:
    if value is None:
        return frozenset(SOURCE_TYPES)
    names = [name.strip().lower() for name in value.split(",") if name.strip()]
    unknown = sorted(set(names).difference(SOURCE_TYPES))
    if unknown:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_CONTENT,
            detail={
                "code": "unsupported_context_source",
                "message": f"Unsupported context source: {', '.join(unknown)}",
            },
        )
    return frozenset(names)  # type: ignore[arg-type]


def _view(result: BusinessContext) -> BusinessContextView:
    return BusinessContextView(
        context_id=result.context_id,
        business_id=str(result.business_id),
        purpose=result.purpose,
        generated_at=result.generated_at,
        token_budget=result.token_budget,
        estimated_tokens=result.estimated_tokens,
        budget_remaining=result.budget_remaining,
        selected_source_types=list(result.selected_source_types),
        sources=[
            ContextSourceView(
                source_type=item.source_type,
                source_reference=item.source_reference,
                source_version=item.source_version,
                authority=item.authority,
                label=item.label,
                updated_at=item.updated_at,
                validity=item.validity,
                selection_status=item.selection_status,
                exclusion_reason=item.exclusion_reason,
                estimated_tokens=item.estimated_tokens,
                content_sha256=item.content_sha256,
                content=item.content,
            )
            for item in result.sources
        ],
        unavailable_sources=result.unavailable_sources,
        context=result.context,
        context_sha256=result.context_sha256,
    )


@router.get("/context", response_model=BusinessContextView)
async def build_context(
    context: Annotated[AuthContext, Depends(require_auth)],
    response: Response,
    purpose: Annotated[str, Query(min_length=1, max_length=80)] = "general",
    token_budget: Annotated[int, Query(ge=256, le=32_768)] = 4096,
    sources: Annotated[str | None, Query(max_length=512)] = None,
    knowledge_query: Annotated[str | None, Query(min_length=1, max_length=500)] = None,
    memory_query: Annotated[str | None, Query(min_length=1, max_length=500)] = None,
) -> BusinessContextView:
    normalized_purpose = purpose.strip().lower()
    if _PURPOSE_PATTERN.fullmatch(normalized_purpose) is None:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_CONTENT,
            detail={
                "code": "invalid_context_purpose",
                "message": "Context purpose must be a lowercase identifier",
            },
        )
    result = await ContextService().build(
        context,
        ContextBuildRequest(
            purpose=normalized_purpose,
            token_budget=token_budget,
            selected_source_types=_source_types(sources),
            knowledge_query=knowledge_query.strip() if knowledge_query else None,
            memory_query=memory_query.strip() if memory_query else None,
        ),
    )
    response.headers["Cache-Control"] = "no-store"
    return _view(result)
